mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala

```diff
@@ -0,0 +1,744 @@
+      step += 1
+      logInfo(s"${sc.appName} adding ${divided.size} new clusterStats at step:${step}")
+
+      if (dividableLeafClusters.isEmpty) {
+        noMoreDividable = true
+      }
+    }
+    // create a map of cluster node with their costs
+    val nodes = createClusterNodes(data, clusterStats)
+    // unpersist RDDs
+    data.unpersist()
+    updatedDataHistory.foreach(_.unpersist())
+
+    // build a cluster tree by Map class which is expressed
+    logInfo(s"Building the cluster tree is started in ${sc.appName}")
+    val root = buildTree(nodes, ROOT_INDEX_KEY, this.k)
+    if (root.isEmpty) {
+      new SparkException("Failed to build a cluster tree from a Map type of clusterStats")
+    }
+
+    // set the elapsed time for training
+    val finishTime = (System.currentTimeMillis() - startTime) / 1000.0
+    logInfo(s"Elapsed Time for ${this.getClass.getSimpleName} Training: ${finishTime} [sec]")
+
+    // make a bisecting kmeans model
+    val model = new BisectingKMeansModel(root.get)
+    val leavesNodes = model.getClusters
+    if (leavesNodes.length < this.k) {
+      logWarning(s"# clusters is less than you want: ${leavesNodes.length} / ${k}")
+    }
+    model
+  }
+}
+
+
+private[clustering] object BisectingKMeans {
+
+  import BisectingClusterStat._
+
+  val ROOT_INDEX_KEY: Long = 1
+
+  /**
+   * Finds the closes cluster's center
+   *
+   * @param metric a distance metric
+   * @param centers centers of the clusters
+   * @param point a target point
+   * @return an index of the array of clusters
+   */
+  def findClosestCenter(metric: (BV[Double], BV[Double]) => Double)
+      (centers: Seq[BV[Double]])(point: BV[Double]): Int = {
+    // get the closest index
+    centers.zipWithIndex.map { case (center, idx) => (metric(center, point), idx)}.minBy(_._1)._2
+  }
+
+  /**
+   * Gets the minimum number of nodes in a tree by the number of leaves
+   *
+   * @param k: the number of leaf nodes
+   */
+  def getMinimumNumNodesInTree(k: Int): Int = {
+    val multiplier = math.ceil(math.log(k) / math.log(2.0))
+    // the calculation is same as `math.pow(2, multiplier)`
+    var numNodes = 2
+    (1 to multiplier.toInt).foreach (i => numNodes = numNodes << 1)
+    numNodes
+  }
+
+  /**
+   * Summarizes data by each cluster as Map
+   *
+   * @param data pairs of point and its cluster index
+   */
+  def summarizeClusters(data: RDD[(Long, BV[Double])]): Map[Long, BisectingClusterStat] = {
+
+    data.mapPartitions { iter =>
+      // calculate the accumulation of the all point in a partition and count the rows
+      val map = mutable.Map.empty[Long, (BV[Double], Double, BV[Double])]
+      iter.foreach { case (idx: Long, point: BV[Double]) =>
+        // get a map value or else get a sparse vector
+        val (sumBV, n, sumOfSquares) = map
+          .getOrElse(idx, (BSV.zeros[Double](point.size), 0.0, BSV.zeros[Double](point.size)))
+        map(idx) = (sumBV + point, n + 1.0, sumOfSquares + (point :* point))
```

Comment:
We don't need a vector for `sumOfSquares` because we only uses its sum to compute in-cluster variance. A single scalar should work.